maketeams3.py

```diff
@@ -258,60 +313,219 @@ def testSwap(teama, playera, teamb, playerb, board):
     for player in players:
         player.setReqMet()
 
-    #WIP for upload format for heltour
+    return {'teams': teams,
+            'players': players,
+            'alternates': alternates,
+            'team_rating_bounds': team_rating_bounds,
+            'alt_rating_bounds': alt_rating_bounds,
+            'alts_split': alts_split}
+
+
+# Reduce variance functions
+
+def intersection(lst1, lst2):
+    return set(lst1).intersection(set(lst2))
+
+
+# Does this swap have a neutral effect on happiness
+def is_neutral_swap(swap):
+    def count_on_team(attr, player, team):
+        n = len(intersection(getattr(player, attr), team.boards))
+        n += len([p for p in team.boards if player in getattr(p, attr)])
+        return n
+
+    count_friends_on_team = partial(count_on_team, 'friends')
+    count_avoids_on_team = partial(count_on_team, 'avoid')
+
+    pa, pb = swap
+    if count_friends_on_team(pa, pa.team) != count_friends_on_team(pa, pb.team)\
+            or count_friends_on_team(pb, pb.team) != count_friends_on_team(pb, pa.team)\
+            or count_avoids_on_team(pa, pa.team) != count_avoids_on_team(pa, pb.team)\
+            or count_avoids_on_team(pb, pb.team) != count_avoids_on_team(pb, pa.team):
+        return False
+
+    return True
+
+
+def get_swaps(teams):
+    num_boards = len(teams[0].boards)
+    boards = [[team.boards[i] for team in teams] for i in range(num_boards)]
+    swaps = [[swap for swap in combinations(board, 2) if is_neutral_swap(swap)] for board in boards]
+    return [swap for board_swaps in swaps for swap in board_swaps]
+
+
+def rating_variance_improvement(league_mean, n_boards, swap):
+    def score(a, b):
+        return variance(league_mean, [a, b])
+
+    pa, pb = swap
+    a_mean = pa.team.getMean()
+    b_mean = pb.team.getMean()
+    initial_score = score(a_mean, b_mean)
+
+    # calculating change in mean if we swapped the players.
+    rating_diff = pb.rating - pa.rating
+    a_mean = a_mean + rating_diff/n_boards
+    b_mean = b_mean - rating_diff/n_boards
+    new_score = score(a_mean, b_mean)
+
+    # lower is better
+    return new_score - initial_score
+
+
+def get_best_swap(swaps, fun):
+    best_swap = min(swaps, key=fun)
+    return best_swap, fun(best_swap)
+
+
+def perform_swap(swap):
+    pa, pb = swap
+    ta = pa.team
+    tb = pb.team
+    board = pa.board
+    ta.changeBoard(board, pb)
+    tb.changeBoard(board, pa)
+
+
+def update_swaps(swaps, swap_performed, teams):
+    pa, pb = swap_performed
+    affected_players = pa.team.boards + pb.team.boards
+    # remove all swaps involving players affected by the swap.
+    swaps = [swap for swap in swaps
+             if not (swap[0] in affected_players or swap[1] in affected_players)]
```

Comment:
nit:  I find this version to be slightly more readable: `swaps = filter(swaps, lambda swap: not intersection(swap, affected_players))`

Reply:
I like the `intersection` idea.  I use list comps and not `filter` because I've been told that's the more pythonic way of doing it.  And I do like it.  list comps are very powerful and flexible.   `map` and `filter` are completely covered.